### ISSUE TITLE
Optionally output package description as comment

### ISSUE THREAD

**Issue by colindean** [Homebrew/homebrew-bundle#322]
I'd like to have a Brewfile that has only the formulae that I've intended to install, no dependencies. Since Homebrew doesn't seem to track† if a formula was installed manually or as a dependency, I'm left to manually edit the output of `brew bundle dump`. It's a slog because I've got more than 350 packages installed but I'll deal.

What could really expedite my discovery of what some formulae are is including as a comment on the line or above it in the Brewfile output. That way, I don't have to do `brew info packagename` for every package I don't recognize, which is a lot since my Homebrew installation dates back a few years. Honestly, I've found that every Brewfile I've manually created for the last couple of months – something I've been dropping into every repo I touch at work – has had a comment describing what the package is or does.

I'm willing to put the time into developing this feature. I envision it as an addition [here](https://github.com/Homebrew/homebrew-bundle/blob/master/lib/bundle/brew_dumper.rb#L123)
 and [here](https://github.com/Homebrew/homebrew-bundle/blob/master/lib/bundle/brew_dumper.rb#L32). It would be enabled through a command line option to the `dump` command, probably something like `--comment` or `--describe`. There's some complexity in that it appears that there are not currently arguments passed down into the dumper.

_† One thing I noted is that maybe this assumption is out of date, given what I see [here](https://github.com/Homebrew/homebrew-bundle/blob/master/lib/bundle/brew_dumper.rb#L25): `f[:installed_on_request?] || !f[:installed_as_dependency?]`_

**Comment by ilovezfs** [comment]
Given you can
```
grep -E '^brew' .Brewfile|awk '{print $2}'|tr -d ','|xargs -n1 brew desc
```
this seems redundant to me.

**Comment by ilovezfs** [comment]
Actually, `-n1` isn't even needed, so that command takes less than a second to run.

**Comment by MikeMcQuaid** [comment]
> Since Homebrew doesn't seem to track† if a formula was installed manually or as a dependency

As noted: we do do this through the attributes you've mentioned.

That said, if this is an optional `dump` argument I'd be fine with this outputting comments. Closing this just because of our Homebrew issue process which is "don't leave open issues that maintainers aren't going to personally implement or actively seek community members to build".

**Comment by ilovezfs** [comment]
One thing that might be applicable here, and generally useful for other similar tasks, would be a `brew bundle list` command so that you could do
```
brew desc $(brew bundle list)
```

**Comment by MikeMcQuaid** [comment]
@ilovezfs Oh, good idea, I like that it could be used to solve a bunch of similar cases. @colindean thoughts?

**Comment by colindean** [comment]
Thanks for the script, @ilovezfs, that greatly accelerates my process. TIL `brew desc`!

Adding `--describe` addresses one of my main use cases that requires some gymnastics with `brew desc` to accomplish: getting that comment in place above each line in the Brewfile. @ilovezfs's script helps my vetting process _now_ but me 6 months from now would cherish the presence of the comment so that I didn't have to pull out that one-liner.

I could probably try my hand at a `list` command, too. It seems the difference between `brew desc $(brew list)` and `brew desc $(brew bundle list)` would be that the latter only lists packages denoted in the active Brewfile. Do I correctly understand that intention, @ilovezfs?

I guess with the `brew bundle list` command in place, "me 6 months from now" could do `brew desc $(brew bundle list)` to see what's there but that doesn't really help when that Brewfile is on Github: I'd need an active Homebrew environment to see what packages are.

**Comment by colindean** [comment]
I've implemented `--describe` a lot faster than I thought I would be able to! It's just a little less useful than I thought it would be: turns out casks don't have a description, only a name; taps don't have information about them other than their name, and `mas` output has just a name. At least I've saved some typing! The tests though… I'm out of rspec practice.

I started plotting out how `brew bundle list` might work. I think I could create a class `Bundle::ListDsl` that overrides the behavior of `Bundle::Dsl#install` to output each entry name instead of installing it. A set of options would output only the dependency type specified: 

    brew bundle list         # outputs everything, kinda messy because we lose dependency type info?
    brew bundle list --brews
    brew bundle list --casks
    brew bundle list --taps
    brew bundle list --mas
    brew bundle list --all   # could alternatively output just Homebrew formulae for the bare command…

Another way to do it would be to refactor `Bundle::Dsl` a bit. Move `Bundle::Dsl#install` to another class, perhaps `Bundle::Installer` to mirror the dumpers structure. This leaves the DSL parsing code in place and lets an instance of `Bundle::Dsl` really be a collection of entries: consumers of it would just act on its `entries`.

Thoughts?
